src/wrap.js

```diff
@@ -11,31 +14,85 @@ import { useImperativeHandle, forwardRef } from 'react';
 // More information on `useImperativeHandle`:
 // <https://reactjs.org/docs/hooks-reference.html#useimperativehandle>
 //
-// functionComponent - The function component you want to test.
-//
 // Example
 //
-//   import { Button } from 'react-native-elements';
-//   import { hook, wrap } from 'cavy';
-//
-//   class MyComponent extends React.Component {
-//     // ...
-//     render() {
-//       const WrappedButton = wrap(Button);
-//
-//       return (
-//         <WrappedButton ref={this.generateTestHook('button')} onPress={}/>
-//       )
-//     }
-//   }
-//   export default hook(MyComponent);
-//
-export default function wrap(functionComponent) {
-  // `forwardRef` accepts a render function that receives our props and ref.
-  return forwardRef((props, ref) => {
-    // It returns the wrapped component after calling `useImperativeHandle`, so
-    // that our ref can be used to call the inner function component's props.
-    useImperativeHandle(ref, () => ({ props }));
-    return functionComponent(props);
-  });
+// import React from 'react';
+// import { Button } from 'react-native-elements';
+// import { useCavy, wrap } from 'cavy';
+//
+// export default () => {
+//   const generateTestHook = useCavy();
+//   const TestableButton = wrap(Button);
+//
+//   return (
+//     <TestableButton ref={this.generateTestHook('button')} onPress={}/>
+//   )
+// };
+//
+//
+//
+// With a component like `Text`:
+//
+// Higher-order component that wraps a component like `Text`, which is neither
+// a React Class nor a Function Component, and returns a React Class with
+// testable props.
+//
+// Example:
+//
+// import React from 'react';
+// import { View, Text } from 'react-native';
+// import { useCavy, wrap } from 'cavy';
+//
+// export default ({ data }) => {
+//   const generateTestHook = useCavy();
+//   const TestableText = wrap(Text);
+//
+//   return (
+//     <View>
+//       <TestableText ref={generateTestHook('title')}>
+//         {data.title}
+//       </TestableText>
+//     </View>
+//   )
+// };
+//
+export default function wrap(Component) {
+  if (typeof Component === 'function' && isNotReactClass(Component)) {
+    // `forwardRef` accepts a render function that receives our props and ref.
+    return forwardRef((props, ref) => {
+      // It returns the wrapped component after calling `useImperativeHandle`, so
+      // that our ref can be used to call the inner function component's props.
+      useImperativeHandle(ref, () => ({ props }));
+      return Component(props);
+    });
+  }
+
+  if (typeof Component == 'object') {
+    class WrapperComponent extends React.Component {
+      render() {
+        return <Component {...this.props} />
+      }
+    }
+    // Wrap the display name for easy debugging.
+    WrapperComponent.displayName = `Wrap(${getDisplayName(Component)})`;
+    return WrapperComponent;
+  }
+
+  const message = "Looks like you're passing a class component into `wrap` - " +
```

Comment:
Are we able to log the name of the component?

Reply:
Should be able to!

Reply:
Ok so this is difficult because component here is a reference to a factory function.

Reply:
OK no worries